keps/sig-storage/20191125-bucket-provisioning.md

```diff
@@ -0,0 +1,667 @@
+1. `conditions`: the latest available observation of this `BucketRequest`. (Phase is not used.)
+1. `Type`: only one type is supported: "BucketAvailable". If the condition's `Status` is true then the bucket is available to be accessed for the `Reason` provided. If `Status` is false then the bucket is unable to be accessed and, again, `Reason` states why.
+1. `Status`: one of true, false or unknown. If not specified then unknown is assumed. A true status indicates the this condition is true at the time of observation.
+1. `Reason`: one of "Bound", "Creating", "Deleting", "Released".
+1. `Message`: optional human readable message containing more details.
+
+#### Bucket
+
+A cluster-scoped custom resource representing the abstraction of a single object store bucket. At a minimum, a `Bucket` instance stores enough identifying information so that drivers can accurately target the backend object store (e.g. needed during a deletion process).  All of the associated bucket classes fields are copied to the `Bucket`. Additionally, endpoint data returned by the driver is copied to the `Bucket` by the sidecar.
+
+There is a 1-to-many relationship between a `Bucket` and a `BucketRequest`, meaning that many `BucketRequest`s can reference the same `Bucket`.
+
+For greenfield, COSI creates the `Bucket` based on values in the `BucketRequest` and `BucketClass`. For brownfield, an admin manually creates the `Bucket` and COSI copies bucket class fields, populates fields returned by the provisioner, and binds the `Bucket` to the `BucketAccess`.
+
+```yaml
+apiVersion: cosi.io/v1alpha1
+kind: Bucket
+metadata:
+  name: [1]
+  labels:
+    cosi.io/provisioner: [2]
+  finalizers:
+    - cosi.io/finalizer [3]
+spec:
+  provisioner: [4]
+  releasePolicy: [5]
+  anonymousAccessMode: [6]
+    - private
+    - publicRead
+    - publicReadWrite
+  bucketClassName: [7]
+  bindings: [8]
+    - "<BucketAccess.name>"
+  protocol: [9]
+    protocolSignature: ""
+    azureBlob:
+      containerName:
+      storageAccount:
+    s3:
+      endpoint:
+      bucketName:
+      region:
+      signatureVersion:
+    gcs:
+      bucketName:
+      privateKeyName:
+      projectId:
+      serviceAccount:
+  allowedNamespaces: [10]
+    - name:
+      uid:
+  parameters: [11]
+status:
+  conditions: [12]
+    - Type: {BucketAvailable} [13]
+      Status: [14]
+      Reason: [15]
+      Message: [16]
+```
+
+1. `name`: When created by COSI, the `Bucket` name is generated in this format: _"bucket-"<bucketRequest.name>"-"<bucketRequest.namespace>_. If an admin creates a `Bucket`, as is necessary for brownfield access, they can use any name.
```

Comment:
Wouldn't `namespace-name` iso `name-namespace` make more sense, e.g. when ordering a `BucketList`, when (tab-)completing in some CLI/UI,...